File: zemiao/75.py
class Solution(object):
    def sortColors(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        if not nums:
          return
        
        l,ll,r = 0,0,len(nums)-1
        #[0,2,0,1,2]
        while ll < r:
          while nums[l]==0:
            l += 1
            if l == r:
              return
          ll = l
          while nums[ll]!=2:
            ll += 1
            if nums[ll]==0:
              nums[l],nums[ll] = nums[ll],nums[l]
              l += 1
            if ll == r:
              return
          while nums[r]==2:
            r -= 1
            if r <= ll:
              return
          nums[ll],nums[r] = nums[r],nums[ll]

File: tnzhou/75.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int leftNext = 0;
        int rightNext = nums.size() - 1;
        while (leftNext < nums.size() && nums[leftNext] == 0)
            leftNext++;
        while (rightNext >= 0 && nums[rightNext] == 2)
            rightNext--;
        for (int i = leftNext; i <= rightNext;) {
            i = max(leftNext, i);
            if (nums[i] == 0) {
                swap(nums[i], nums[leftNext]);
                i++;
                leftNext++;
            } else if (nums[i] == 2) {
                swap(nums[i], nums[rightNext]);
                rightNext--;
            } else {
                i++;
            }
        }
    }
};
